File: js/src/expression.ts
import * as yaml from "yaml"
import { JoinError } from "./error"
import { applyInMatcher } from "./in"
import { ParserData } from "./lidy"
import { applyListMatcher } from "./list"
import { applyMapMatcher } from "./map"
import { applyOneOfMatcher } from "./oneOf"
import { applyRangeMatcher } from "./range"
import { applyRegexMatcher } from "./regex"
import { Result } from "./result"
import { applyRule } from "./rule"
import { applySizeCheck } from "./size"

export function applyExpression(
  parserData: ParserData,
  schema: yaml.YAMLMap<any, any>,
  content: yaml.Node,
): Result<any> {
  if (yaml.isScalar<any>(schema)) {
    const { value } = schema
    if (typeof value !== "string") {
      throw new Error(
        `encountered a value: ${value} (type: ${typeof value}) where an expression was expected`,
      )
    }
    return applyRule(parserData, value, content)
  }

  // Else, the expression must be a mapping
  if (!yaml.isMap<any, any>(schema)) {
    throw new Error(
      "Lidy expressions must be strings (rule names) or mappings (checkers)",
    )
  }

  let _map,
    _mapFacultative,
    _mapOf,
    _merge,
    _list,
    _listFacultative,
    _listOf,
    _min,
    _max,
    _nb

  for (const { key, value } of schema.items) {
    switch (key.value) {
      case "_regex":
        return applyRegexMatcher(parserData, value, content)
      case "_in":
        return applyInMatcher(parserData, value, content)
      case "_range":
        return applyRangeMatcher(parserData, value, content)
      case "_oneOf":
        return applyOneOfMatcher(parserData, value, content)
      case "_map":
        _map = value
        break
      case "_mapFacultative":
        _mapFacultative = value
        break
      case "_mapOf":
        _mapOf = value
        break
      case "_merge":
        _merge = value
        break
      case "_list":
        _list = value
        break
      case "_listFacultative":
        _listFacultative = value
        break
      case "_listOf":
        _listOf = value
        break
      case "_min":
        _min = value
        break
      case "_max":
        _max = value
        break
      case "_nb":
        _nb = value
        break
      default:
        throw new Error(`Unknown keyword found in matcher: '${key.value}'`)
    }
  }

  let joinError = new JoinError()
  if (_min || _max || _nb) {
    joinError.add(applySizeCheck(parserData, content as any, _min, _max, _nb))
  }
  let result: Result<any> | undefined
  if (_map || _mapFacultative || _mapOf || _merge) {
    try {
      result = applyMapMatcher(
        parserData,
        _map,
        _mapFacultative,
        _mapOf,
        _merge,
        content,
      )
    } catch (e) {
      joinError.add(e)
    }
  }
  if (_list || _listFacultative || _listOf) {
    try {
      result = applyListMatcher(
        parserData,
        _list,
        _listFacultative,
        _listOf,
        content,
      )
    } catch (e) {
      joinError.add(e)
    }
  }
  joinError.throw()
  if (result) {
    return result
  } else {
    throw new Error("No keyword found in matcher")
  }
}

File: rust/src/expression.rs
use lidy__yaml::{Yaml, YamlData};

use crate::error::{AnyBoxedError, JoinError, SimpleError};
use crate::in_::apply_in_matcher;
use crate::list::apply_list_matcher;
use crate::map::apply_map_matcher;
use crate::one_of::apply_one_of_matcher;
use crate::parser::Parser;
use crate::range::apply_range_matcher;
use crate::regex::apply_regex_matcher;
use crate::result::LidyResult;
use crate::rule::apply_rule;
use crate::size::apply_size_check;

pub fn apply_expression<TV>(
    parser: &mut Parser<TV>,
    schema: &Yaml,
    content: &Yaml,
) -> Result<LidyResult<TV>, AnyBoxedError>
where
{
    match &schema.data {
        YamlData::String(value) => apply_rule(parser, value, content),
        YamlData::Mapping(mapping) => {
            let mut map = None;
            let mut map_facultative = None;
            let mut map_of = None;
            let mut merge = None;
            let mut list = None;
            let mut list_facultative = None;
            let mut list_of = None;
            let mut min = None;
            let mut max = None;
            let mut nb = None;

            for (key, value) in mapping {
                if let YamlData::String(key_str) = &key.data {
                    match key_str.as_str() {
                        "_regex" => return apply_regex_matcher(parser, value, content),
                        "_in" => return apply_in_matcher(parser, value, content),
                        "_range" => return apply_range_matcher(parser, value, content),
                        "_oneOf" => return apply_one_of_matcher(parser, value, content),
                        "_map" => map = Some(value),
                        "_mapFacultative" => map_facultative = Some(value),
                        "_mapOf" => map_of = Some(value),
                        "_merge" => merge = Some(value),
                        "_list" => list = Some(value),
                        "_listFacultative" => list_facultative = Some(value),
                        "_listOf" => list_of = Some(value),
                        "_min" => min = Some(value),
                        "_max" => max = Some(value),
                        "_nb" => nb = Some(value),
                        _ => {
                            return Err(SimpleError::from_message(
                                format!("Unknown keyword found in matcher: '{key_str}'").into(),
                            )
                            .into())
                        }
                    }
                }
            }

            let mut join_error = JoinError::default();

            let is_mapping =
                map.is_some() || map_facultative.is_some() || map_of.is_some() || merge.is_some();
            let is_list = list.is_some() || list_facultative.is_some() || list_of.is_some();

            if is_mapping && is_list {
                return Err(SimpleError::from_check(
                    "_(map*|list*)",
                    "Cannot apply _map and _list at the same time",
                    schema,
                )
                .into());
            }

            let result = if is_mapping {
                Some(apply_map_matcher(
                    parser,
                    map,
                    map_facultative,
                    map_of,
                    merge,
                    content,
                )?)
            } else if is_list {
                Some(apply_list_matcher(
                    parser,
                    list,
                    list_facultative,
                    list_of,
                    content,
                )?)
            } else {
                return Err(SimpleError::from_check(
                    "_(map*|list*)",
                    "no keyword found in matcher",
                    schema,
                )
                .into());
            };

            if min.is_some() || max.is_some() || nb.is_some() {
                if let Some(error) = apply_size_check(content, min, max, nb) {
                    join_error.add(error);
                }
            }

            join_error.into_result()?;

            Ok(result.unwrap())
        }
        _ => panic!("Lidy expressions must be strings (rule names) or mappings (checkers)"),
    }
}
